doc/developer/platform/formalism.md

```diff
@@ -765,4 +760,4 @@ These lower layers should provide similar "transactional" interfaces that valida
 
 As Materialize runs, the Adapter may see fit to "allow compaction" of collections Materialize maintains.
 It does so by downgrading its held read capabilities for the collection (identifier by `GlobalId`).
-Downgraded capabilities restrict the ability of Adapter to form commands to Storage and Compute, and may force the timestamp oracle's timestamps forward. Generally, the Adapter should not downgrade its read capabilities past the timestamp oracle's current timestamp, thereby avoiding this constraint.
+Downgraded capabilities restrict the ability of Adapter to form commands to Storage and Compute, and may force the timestamp oracle's timestamps forward.
```

Comment:
Or delay query response, which could be a better reaction to tainting the timestamp oracle.

Reply:
I added "or delay query response" at the end.